ariadne/src/Ariadne/Wallet/Backend/Tx.hs

```diff
@@ -74,22 +77,55 @@ sendTx ::
     -> InputSelectionPolicy
     -> NonEmpty TxOut
     -> IO TxId
-sendTx acidDb WalletFace {..} CardanoFace {..} walletSelRef printAction pp walletRef _accRefs isp outs = do
+sendTx acidDb WalletFace {..} CardanoFace {..} walletSelRef printAction pp walletRef accRefs isp outs = do
     let Nat runCardanoMode = cardanoRunCardanoMode
     walletDb <- query acidDb Snapshot
     let wallets = walletDb ^. dbHdWallets
-    walletRootId <- resolveWalletRef walletSelRef walletRef walletDb
-    runCardanoMode $ sendTxDo wallets walletRootId =<< cardanoGetDiffusion
+    (walletRootId, walletAccounts) <-
+        resolveWalletRefThenRead walletSelRef walletRef
+        walletDb readAccountsByRootId
+    accountIds <- resolveAccountRefs walletRootId walletAccounts
+    let filteredAccounts :: IxSet HdAccount
+        filteredAccounts = maybe identity filterAccounts accountIds walletAccounts
+
+        filterAccounts :: NonEmpty HdAccountId -> IxSet HdAccount -> IxSet HdAccount
+        filterAccounts ids accounts = accounts @+ toList ids
+
+    runCardanoMode $
+        sendTxDo wallets walletRootId filteredAccounts =<< cardanoGetDiffusion
   where
-    sendTxDo :: HdWallets -> HdRootId -> Diffusion CardanoMode -> CardanoMode TxId
-    sendTxDo wallets walletRootId diffusion = do
-        let
-            walletAccounts :: IxSet HdAccount
-            walletAccounts = fromRight
-                -- Resolved walletRootId always exists.
-                (error "SendTx: Unknown walletRootId")
-                (readAccountsByRootId walletRootId wallets)
-
+    -- Returns list of accounts which can be used.
+    -- 'Nothing' means all accounts can be used.
+    resolveAccountRefs ::
```

Comment:
I don't want to move it inside, because then I'll have to return too much stuff (`walletRootId` and `walletAccounts`) are also needed outside. I can rename this function to e. g. `getSuitableAccounts`.

Reply:
`getSuitableAccounts` sounds good to me.